Action group button component module (#15)
evargast/signature-generator: src/components/ActionGroupButtonComponent/README.md

## Notes about the component

> OK so some notes on the made component the process and the use of the stroy book and how everything is wired up.

Reviewer: ```suggestion
OK so some notes on the component, the process and the use of the storybook and how everything is wired up.
```

## UI

First the integration of the UI. This was done though the return statement of the FC where all the html tags and JSX components are, here the pertinent keys that are used for such elements are defined and thus be used in the component behaviour logic. e.g. 'isBold' key in the react spectrum components.

## Pertinent Properties

> First once the user interface has been defined the needed interface values need to be also defined, with the relevant types, the _?_ if the value is to be optional, if it is left without asignment its undefined. as such:

Reviewer: ```suggestion
First once the user interface has been defined the needed interface values need to be also defined, with the relevant types, the `?` if the value is to be optional, if it is left without assignment its undefined. as such:
```

```typescript
interface Props {
    isBold?: boolean;
    isItalics?: boolean;
}
```

### Note

This are the components to be passed both in the FC:

```typescript
const ActionGroupButtonComponent: FC<Props> = ({ isBold, isItalics }) => {};
```

that are then accessible in the storybook expanded arguments.

```typescript

> [...args];

Reviewer: ```suggestion
{...args};
```

```

### Setting the values.

The initial request was to use an object yet as the problem, but after some refinement instead of an object the next best implementation was for the use of given useState hook provided by the Adobe documentation:

```typescript
const [selected, setSelected] = React.useState<Selection>(new Set(["list"]));
```

> Then due to the fact that we needed to set the interface values, and this was not done by the default hook the implementation of an complex hook was implemented, with the explicit use of the anonimous funtion, like so:

Reviewer: ```suggestion
Then due to the fact that we needed to set the interface values, and this was not done by the default hook, a complex hook was implemented, with the explicit use of the anonymous function, like so:
```

```typescript
const [selected, setSelected] = React.useState<Selection>(() => {});
```

By using the complex implementation of the useState we then create an array within it where the keys are pushed depending on the interface values. the implementation becomes as such with the logic:

```typescript
const initialValue: string[] = [];

if (isBold) {
    initialValue.push("bold");
}

if (isItalics) {
    initialValue.push("italics");
}

return new Set(initialValue);
```

All of this within the complex useState hook.

> Very important to return the expected type, in this case Selection expects a Set so a new Set is made with the elements of the local array that have been pushed.

Reviewer: ```suggestion
Very important to return the expected type, in this case Selection expects a `Set` so a new Set is created with the local array that has been modified within the anonymous function.
```

> At this point the setSelected is called every time onSelectionChange prop is called this updating the values of the array. And being tracked by the selectedKeys prop of the Action Group component. With this impementation we have controlled by having the selectedKeys prop having the selected interface values.

Reviewer: ```suggestion
At this point the setSelected is called every time `onSelectionChange` prop is called thus updating the values of the array. And being tracked by the `selectedKeys` prop of the Action Group component. With this implementation we have a controlled  ActionGroup by passing the `selectedKeys` prop with the selected values.
```

## Storybook setup

> Within the storybook the props from the interface are passed with the ...args, thus the Storybook has access to the interface references that have been also passe into the FC.

Reviewer: ```suggestion
Within the storybook the props from the interface are passed with the `...args`, thus the Storybook has access to the props/arguments that have been also passed into the Component.
```

> We then make the specific stories that are all the branches of possible state combinations in this case, all switches off, all on and one or the other. this follows the following syntax e.g.:

Reviewer: ```suggestion
We then make the specific stories that are all the bound to the sample story template (using the `bind()` method). Each story illustrates a different possible state, in this case,
1. on and on
2. off and off
3. on and off  
4. off and on

 This follows the following syntax e.g.:
```
